File: app/scoring.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from .config import UserProfile
from .models import Decision, JobScore
from .normalize import normalize_text


OFFICIAL_ATS = {"greenhouse", "lever", "ashby", "smartrecruiters", "generic_careers"}

EXTRA_FIT_TERMS = [
    "producer",
    "senior producer",
    "development producer",
    "creative producer",
    "content producer",
    "content",
    "creative",
    "development",
    "branded content",
    "partnerships",
    "partner manager",
    "content strategy",
    "creative strategy",
    "editorial",
    "media",
    "communications",
    "public relations",
    "campaign",
    "client",
    "account manager",
    "commercial",
    "growth",
    "gtm",
    "customer success",
    "entertainment",
    "television",
    "tv",
    "video",
    "factual",
    "unscripted",
    "research",
]

LOW_FIT_TERMS = [
    "software engineer",
    "frontend engineer",
    "backend engineer",
    "full stack engineer",
    "machine learning engineer",
    "data engineer",
    "legal counsel",
    "lawyer",
    "retail",
    "store",
    "genius bar",
    "warehouse",
    "driver",
    "internship",
    "apprenticeship",
]


def contains_any(text: str, phrases: list[str]) -> list[str]:
    normalized = normalize_text(text)
    return [phrase for phrase in phrases if normalize_text(phrase) in normalized]


def score_job(row: Mapping[str, Any], profile: UserProfile) -> JobScore:
    title = str(row.get("title") or "")
    description = str(row.get("description") or "")
    salary_text = str(row.get("salary_text") or "")
    combined = " ".join([title, description, salary_text])
    normalized_combined = normalize_text(combined)

    ats_type = str(row.get("ats_type") or "").lower()
    source = str(row.get("source") or "").lower()
    official_posting_score = 10 if ats_type in OFFICIAL_ATS and source in OFFICIAL_ATS else 0

    sponsor_status = str(row.get("sponsor_status") or "")
    sponsor_rating = str(row.get("sponsor_rating") or "")
    sponsor_routes = str(row.get("sponsor_routes") or "")
    sponsor_score = 0
    if sponsor_status == "matched":
        sponsor_score = 8
        if "a" in sponsor_rating.casefold():
            sponsor_score = 10
        if "skilled worker" in sponsor_routes.casefold():
            sponsor_score = min(10, sponsor_score + 1)

    min_salary = row.get("min_salary")
    max_salary = row.get("max_salary")
    salary_score = 5
    if min_salary is not None or max_salary is not None:
        best_salary = int(max_salary or min_salary or 0)
        floor_salary = int(min_salary or max_salary or 0)
        if best_salary < profile.minimum_salary_general:
            salary_score = 0
        elif floor_salary >= profile.minimum_salary_general:
            salary_score = 10
        else:
            salary_score = 7

    lane_hits = contains_any(combined, profile.preferred_lanes)
    seniority_hits = contains_any(combined, profile.seniority_keywords)
    extra_fit_hits = contains_any(combined, EXTRA_FIT_TERMS)
    low_fit_hits = contains_any(combined, LOW_FIT_TERMS)
    cv_fit_score = min(10, len(lane_hits) * 3 + len(seniority_hits) * 2 + len(extra_fit_hits))
    if low_fit_hits:
        cv_fit_score = max(0, cv_fit_score - 5)

    soc_terms = [
        "producer",
        "communications manager",
        "marketing manager",
        "public relations",
        "research",
        "creative director",
        "content strategist",
        "partnerships manager",
        "account manager",
        "customer success",
        "commercial manager",
    ]
    soc_hits = contains_any(combined, soc_terms)
    soc_score = min(10, 4 + len(soc_hits) * 2) if soc_hits else 3

    hard_reject_hits = contains_any(combined, profile.hard_reject_phrases)
    strong_sponsorship_hits = contains_any(combined, profile.strong_sponsorship_phrases)
    sponsorship_language_score = 9 if strong_sponsorship_hits else 5
    if hard_reject_hits:
        sponsorship_language_score = 0

    total_score = (
        official_posting_score
        + sponsor_score
        + salary_score
        + soc_score
        + cv_fit_score
        + sponsorship_language_score
    )

    decision = Decision.REJECT
    rejection_reason: str | None = "Insufficient match"
    if sponsor_score == 0:
        rejection_reason = "Company not matched to licensed sponsor register"
    elif official_posting_score < 8:
        rejection_reason = "Not verified as an official company or ATS posting"
    elif salary_score == 0:
        rejection_reason = f"Salary appears below GBP {profile.minimum_salary_general}"
    elif hard_reject_hits:
        rejection_reason = "Posting contains hard visa sponsorship rejection language"
    elif low_fit_hits and cv_fit_score < 5:
        rejection_reason = "Role appears outside target lanes"
    elif sponsor_score >= 7 and salary_score >= 6 and cv_fit_score >= 7 and official_posting_score >= 8:
        decision = Decision.APPLY
        rejection_reason = None
    elif sponsor_score >= 7 and official_posting_score >= 8 and cv_fit_score >= 5 and salary_score >= 5:
        decision = Decision.HOLD
        rejection_reason = "Sponsor matched and CV fit is plausible, but salary or sponsorship language needs human verification"
    elif sponsor_score >= 7 and official_posting_score >= 8 and cv_fit_score >= 4 and any(
        term in normalized_combined for term in ("producer", "content", "creative", "partnership", "communications", "media")
    ):
        decision = Decision.HOLD
        rejection_reason = "Potentially relevant sponsor role; needs human review"

    return JobScore(
        job_id=int(row["id"]),
        official_posting_score=official_posting_score,
        sponsor_score=sponsor_score,
        salary_score=salary_score,
        soc_score=soc_score,
        cv_fit_score=cv_fit_score,
        sponsorship_language_score=sponsorship_language_score,
        total_score=total_score,
        decision=decision,
        rejection_reason=rejection_reason,
        evidence_json={
            "lane_hits": lane_hits,
            "seniority_hits": seniority_hits,
            "extra_fit_hits": extra_fit_hits,
            "low_fit_hits": low_fit_hits,
            "soc_hits": soc_hits,
            "hard_reject_hits": hard_reject_hits,
            "strong_sponsorship_hits": strong_sponsorship_hits,
            "salary_text": salary_text,
            "sponsor_rating": sponsor_rating,
            "sponsor_routes": sponsor_routes,
        },
    )

File: lib/scoring.ts
import type { SponsorStatus } from '@prisma/client';

export type ScoringInput = {
  title: string;
  companyName: string;
  location?: string | null;
  sourceUrl: string;
  officialApplicationUrl?: string | null;
  jdText: string;
  salaryText?: string | null;
  sponsorStatus?: SponsorStatus | 'UNKNOWN' | 'LICENSED' | 'EXPLICIT' | 'NONE';
  sponsorshipEvidence?: string | null;
};

export type ScoredJob = {
  detectedSeniority: string;
  detectedRoleFamily: string;
  salarySocScore: number;
  cvFitScore: number;
  overallScore: number;
  status: 'REJECTED' | 'REVIEW' | 'READY_TO_APPLY';
  rejectionReasons: Array<{ code: string; message: string }>;
};

const positiveKeywords = [
  'producer',
  'development',
  'content',
  'creative',
  'partnerships',
  'media',
  'communications',
  'strategy',
  'branded content',
  'entertainment',
  'factual',
  'unscripted',
  'video',
  'tv',
  'television',
  'comedy',
  'gtm',
  'customer success',
  'partnerships manager',
  'content strategy',
  'creator',
  'campaign',
  'editorial'
];

const strongKeywords = [
  'development producer',
  'creative producer',
  'content producer',
  'senior producer',
  'branded content',
  'factual entertainment',
  'unscripted',
  'television',
  'media partnerships',
  'content strategy'
];

const negativeKeywords = [
  'intern',
  'internship',
  'volunteer',
  'unpaid',
  'junior assistant',
  'admin assistant',
  'receptionist',
  'warehouse',
  'retail assistant',
  'barista',
  'software engineer',
  'backend engineer',
  'frontend engineer',
  'full stack engineer'
];

function normalise(value?: string | null) {
  return (value ?? '').toLowerCase().replace(/\s+/g, ' ').trim();
}

function combined(job: ScoringInput) {
  return normalise(`${job.title} ${job.companyName} ${job.location ?? ''} ${job.jdText} ${job.salaryText ?? ''}`);
}

function containsAny(text: string, terms: string[]) {
  return terms.some((term) => text.includes(term));
}

export function detectSeniority(job: ScoringInput) {
  const text = combined(job);
  if (/\b(chief|vp|vice president|svp|director|head of)\b/.test(text)) return 'senior_leadership';
  if (/\b(senior|lead|principal|manager)\b/.test(text)) return 'senior_or_manager';
  if (/\b(associate|executive|coordinator)\b/.test(text)) return 'mid_or_entry';
  if (/\b(intern|assistant|junior)\b/.test(text)) return 'junior';
  return 'unknown';
}

export function detectRoleFamily(job: ScoringInput) {
  const text = combined(job);
  if (/\b(unscripted|television|tv|factual|entertainment|producer|development producer)\b/.test(text)) {
    return 'tv_content_production';
  }
  if (/\b(branded content|creative strategy|campaign|creative producer|content strategy)\b/.test(text)) {
    return 'branded_content_strategy';
  }
  if (/\b(partnership|business development|gtm|customer success|account manager)\b/.test(text)) {
    return 'partnerships_gtm';
  }
  if (/\b(communications|pr|media relations|editorial)\b/.test(text)) return 'media_comms';
  if (/\b(engineer|developer|software|data scientist)\b/.test(text)) return 'technical_low_fit';
  return 'unknown';
}

export function scoreCvFit(job: ScoringInput) {
  const text = combined(job);
  let score = 0;

  for (const keyword of positiveKeywords) {
    if (text.includes(keyword)) score += 5;
  }
  for (const keyword of strongKeywords) {
    if (text.includes(keyword)) score += 10;
  }
  if (detectSeniority(job) === 'senior_or_manager') score += 12;
  if (detectSeniority(job) === 'senior_leadership') score += 8;
  if (detectRoleFamily(job) === 'technical_low_fit') score -= 40;
  if (containsAny(text, negativeKeywords)) score -= 35;

  return Math.max(0, Math.min(100, score));
}

export function scoreSalarySocViability(job: ScoringInput) {
  const text = combined(job);
  const salary = normalise(job.salaryText);
  let score = 20;

  const salaryNumbers = [...salary.matchAll(/£?\s?(\d{2,3})(?:,?000|k)/g)].map((m) => Number(m[1]) * 1000);
  const maxSalary = salaryNumbers.length ? Math.max(...salaryNumbers) : null;

  if (maxSalary && maxSalary >= 45000) score += 55;
  else if (maxSalary && maxSalary >= 41700) score += 45;
  else if (maxSalary && maxSalary >= 35000) score += 10;
  else if (maxSalary && maxSalary < 35000) score -= 35;

  if (/\b(senior|lead|manager|director|head of|principal)\b/.test(text)) score += 20;
  if (/\b(part time|temporary|zero hours|internship|apprentice|volunteer|unpaid)\b/.test(text)) score -= 45;
  if (/\b(fixed term|ftc|contract)\b/.test(text) && !/sponsor|visa|skilled worker/.test(text)) score -= 25;

  return Math.max(0, Math.min(100, score));
}

export function hasSponsorEvidence(job: ScoringInput) {
  const text = combined(job);
  return (
    job.sponsorStatus === 'LICENSED' ||
    job.sponsorStatus === 'EXPLICIT' ||
    Boolean(job.sponsorshipEvidence?.trim()) ||
    /\b(sponsorship|visa sponsorship|skilled worker|certificate of sponsorship|cos)\b/.test(text)
  );
}

export function getRejectionReasons(job: ScoringInput) {
  const text = combined(job);
  const reasons: Array<{ code: string; message: string }> = [];

  if (!job.officialApplicationUrl && !job.sourceUrl) {
    reasons.push({ code: 'NO_OFFICIAL_URL', message: 'No official company or ATS application URL is available.' });
  }
  if (!hasSponsorEvidence(job)) {
    reasons.push({ code: 'NO_SPONSOR_EVIDENCE', message: 'No licensed-sponsor or explicit sponsorship evidence is attached.' });
  }
  if (scoreSalarySocViability(job) < 45) {
    reasons.push({ code: 'SALARY_SOC_RISK', message: 'Salary or seniority appears weak for Skilled Worker viability.' });
  }
  if (scoreCvFit(job) < 45) {
    reasons.push({ code: 'LOW_CV_FIT', message: 'The role does not strongly match Karan’s TV/content/partnerships profile.' });
  }
  if (containsAny(text, negativeKeywords)) {
    reasons.push({ code: 'NEGATIVE_ROLE_SIGNAL', message: 'The role contains junior, unpaid, assistant, retail, warehouse, or pure engineering signals.' });
  }
  if (/\b(fixed term|ftc|contract)\b/.test(text) && !/sponsor|visa|skilled worker/.test(text)) {
    reasons.push({ code: 'FTC_WITHOUT_SPONSORSHIP', message: 'FTC/contract role without explicit sponsorship evidence.' });
  }

  return reasons;
}

export function scoreJob(job: ScoringInput): ScoredJob {
  const cvFitScore = scoreCvFit(job);
  const salarySocScore = scoreSalarySocViability(job);
  const detectedSeniority = detectSeniority(job);
  const detectedRoleFamily = detectRoleFamily(job);
  const rejectionReasons = getRejectionReasons(job);
  const sponsorBoost = hasSponsorEvidence(job) ? 15 : 0;
  const overallScore = Math.max(0, Math.min(100, Math.round(cvFitScore * 0.5 + salarySocScore * 0.35 + sponsorBoost)));
  const status = rejectionReasons.length > 0 ? 'REJECTED' : overallScore >= 65 ? 'READY_TO_APPLY' : 'REVIEW';

  return {
    detectedSeniority,
    detectedRoleFamily,
    salarySocScore,
    cvFitScore,
    overallScore,
    status,
    rejectionReasons
  };
}
